versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java

```diff
@@ -0,0 +1,401 @@
+    return toProto(entry).getSerializedSize();
+  }
+
+  @Override
+  protected Stream<KeyListEntity> fetchKeyLists(
+      NonTransactionalOperationContext ctx, List<Hash> keyListsIds) {
+    return fetchMappedPage(
+        client.getKeyLists(),
+        keyListsIds,
+        document -> {
+          Hash hash = idAsHash(document);
+          KeyList keyList = protoToKeyList(data(document));
+          return KeyListEntity.of(hash, keyList);
+        })
+        .stream();
+  }
+
+  @Override
+  protected void writeIndividualCommit(NonTransactionalOperationContext ctx, CommitLogEntry entry)
+      throws ReferenceConflictException {
+    insert(client.getCommitLog(), entry.getHash(), toProto(entry).toByteArray());
+  }
+
+  @Override
+  protected void writeMultipleCommits(
+      NonTransactionalOperationContext ctx, List<CommitLogEntry> entries)
+      throws ReferenceConflictException {
+    List<Document> docs =
+        entries.stream()
+            .map(e -> toDoc(e.getHash(), toProto(e).toByteArray()))
+            .collect(Collectors.toList());
+    insert(client.getCommitLog(), docs);
+  }
+
+  @Override
+  protected void writeKeyListEntities(
+      NonTransactionalOperationContext ctx, List<KeyListEntity> newKeyListEntities) {
+    for (KeyListEntity keyList : newKeyListEntities) {
+      try {
+        insert(client.getKeyLists(), keyList.getId(), toProto(keyList.getKeys()).toByteArray());
+      } catch (ReferenceConflictException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+  }
+
+  @Override
+  protected void writeGlobalCommit(NonTransactionalOperationContext ctx, GlobalStateLogEntry entry)
+      throws ReferenceConflictException {
+    String id = toId(Hash.of(entry.getId()));
+    insert(client.getGlobalLog(), toDoc(id, entry.toByteArray()));
+  }
+
+  @Override
+  protected void unsafeWriteGlobalPointer(
+      NonTransactionalOperationContext ctx, GlobalStatePointer pointer) {
+    Document doc = toDoc(pointer);
+
+    UpdateResult result =
+        client
+            .getGlobalPointers()
+            .updateOne(
+                Filters.eq((Object) globalPointerKey),
+                new Document("$set", doc),
+                new UpdateOptions().upsert(true));
+
+    if (!result.wasAcknowledged()) {
+      throw new IllegalStateException(
+          "Unacknowledged write to " + client.getGlobalPointers().getNamespace());
+    }
+  }
+
+  @Override
+  protected boolean globalPointerCas(
+      NonTransactionalOperationContext ctx,
+      GlobalStatePointer expected,
+      GlobalStatePointer newPointer) {
+    Document doc = toDoc(newPointer);
+    byte[] expectedGlobalId = expected.getGlobalId().toByteArray();
+
+    UpdateResult result =
+        client
+            .getGlobalPointers()
+            .updateOne(
+                Filters.and(Filters.eq(globalPointerKey), Filters.eq("globalId", expectedGlobalId)),
+                new Document("$set", doc));
+
+    return result.wasAcknowledged()
```

Comment:
This CAS-operation LGTM from my understanding of MongoDB. But would be good if someone else could confirm.